src/init/application_feature.rs

```diff
@@ -0,0 +1,76 @@
+use std::any::Any;
+use crate::init::{device, instance};
+use crate::rosella::InstanceContext;
+use crate::util::extensions::ExtensionFunctionSet;
+use crate::UUID;
+
+
+/// Common functions requires by all features
+pub trait FeatureBase {
+    fn as_any(&self) -> &dyn Any;
+
+    fn as_any_mut(&mut self) -> &mut dyn Any;
+}
+
+/// Represents the result of a init operation of a feature
+pub enum InitResult {
+    /// Indicates that the feature is supported and can be enabled
+    Ok,
+
+    /// Indicates that the feature is not supported and must be disabled
+    Disable,
+}
+
+/// A feature that controls instance creation
+///
+/// See [`crate::init::instance`] for more information.
+pub trait ApplicationInstanceFeature : FeatureBase {
+
+    /// Tests if the feature is supported
+    fn init(&mut self, features: &mut dyn FeatureAccess, info: &instance::InstanceInfo) -> InitResult;
+
+    /// Configures the instance
+    fn enable(&mut self, features: &mut dyn FeatureAccess, info: &instance::InstanceInfo, config: &mut instance::InstanceConfigurator);
+
+    /// Performs any necessary post creation steps and generates the data that is sent back to the application
+    fn finish(&mut self, _: &ash::Instance, _: &ExtensionFunctionSet) -> Option<Box<dyn Any>> {
+        None
+    }
+}
+
+/// A object that can generate [`ApplicationDeviceFeature`] instances used during the device
+/// creation process.
+///
+/// See [`crate::init::device`] for more information.
+pub trait ApplicationDeviceFeatureGenerator {
+
+    /// Creates a new instance
+    fn make_instance(&self) -> Box<dyn ApplicationDeviceFeature>;
+}
+
+/// A feature that controls device creation
+///
+/// See [`crate::init::device`] for more information.
+pub trait ApplicationDeviceFeature: Send + FeatureBase {
+
+    /// Tests if the feature is supported
+    fn init(&mut self, features: &mut dyn FeatureAccess, info: &device::DeviceInfo) -> InitResult;
+
+    /// Configures the device
+    fn enable(&mut self, features: &mut dyn FeatureAccess, info: &device::DeviceInfo, config: &mut device::DeviceConfigurator);
+
+    fn finish(&mut self, _: &InstanceContext, _: &ash::Device, _: &ExtensionFunctionSet) -> Option<Box<dyn Any>> {
+        None
+    }
+}
+
+/// Trait used by features to access their dependencies
+pub trait FeatureAccess {
+    fn get(&self, feature: &UUID) -> Option<&dyn Any>;
```

Comment:
What's this meant to return? Shouldn't this be more concrete with something like
```rs
fn get<F: ApplicationDeviceFeature>(&self, feature: &UUID) -> Option<&F::Data>;
```
where Data would be an associated type

Reply:
Couldn't `ApplicationDeviceFeature` have an associated constant UUID to remove the parameter too?

Reply:
The data that is returned could be anything the feature wants to return. A set of queues, some information about how a feature was enabled or nothing at all. I can be more specific than Any because it is completely up the the feature to determine what it needs to return.
I the application requires that data it is up to it to know what type that is (for example through some associated constant). I cannot force the use of an associated constant because that would make it impossible to use dyn on the trait object.

As for the uuid the issue is that some features need to be able to be added multiple times. For example you should be able to add multiple surface features. If the uuid is a associated constant that is impossible because uniqueness would be tied to the type not the instance. I added utility macros to generate features who do not need that flexibility which does use a associated constant.